main/src/hadd.cxx

```diff
@@ -90,284 +102,473 @@
   \note By default histograms are added. However hadd does not support the case where
          histograms have their bit TH1::kIsAverage set.
 
-  \authors Rene Brun, Dirk Geppert, Sven A. Schmidt, Toby Burnett
+  \authors Rene Brun, Dirk Geppert, Sven A. Schmidt, Toby Burnett, G. Parolini
 */
 #include "Compression.h"
-#include <ROOT/RConfig.hxx>
-#include "ROOT/TIOFeatures.hxx"
+#include "TClass.h"
 #include "TFile.h"
+#include "TFileMerger.h"
 #include "THashList.h"
 #include "TKey.h"
-#include "TClass.h"
 #include "TSystem.h"
 #include "TUUID.h"
-#include "ROOT/StringConv.hxx"
-#include "snprintf.h"
 
-#include <string>
-#include <iostream>
-#include <fstream>
-#include <cstdlib>
+#include <ROOT/RConfig.hxx>
+#include <ROOT/StringConv.hxx>
+#include <ROOT/TIOFeatures.hxx>
+
+#include "haddCommandLineOptionsHelp.h"
+
 #include <climits>
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <optional>
 #include <sstream>
-#include "haddCommandLineOptionsHelp.h"
+#include <string>
 
-#include "TFileMerger.h"
 #ifndef R__WIN32
 #include "ROOT/TProcessExecutor.hxx"
 #endif
 
 ////////////////////////////////////////////////////////////////////////////////
 
-int main( int argc, char **argv )
+inline std::ostream &Err() {
+   std::cerr << "Error in <hadd>: ";
+   return std::cerr;
+}
+
+inline std::ostream &Warn() {
+   std::cerr << "Warning in <hadd>: ";
+   return std::cerr;
+}
+
+inline std::ostream &Info() {
+   std::cerr << "Info in <hadd>: ";
+   return std::cerr;
+}
+
+using IntFlag_t = uint32_t;
+
+struct HAddArgs {
+   bool fNoTrees;
+   bool fAppend;
+   bool fForce;
+   bool fSkipErrors;
+   bool fReoptimize;
+   bool fDebug;
+   bool fKeepCompressionAsIs;
+   bool fUseFirstInputCompression;
+
+   std::optional<std::string> fWorkingDir;
+   std::optional<IntFlag_t> fNProcesses;
+   std::optional<TString> fCacheSize;
+   std::optional<ROOT::TIOFeatures> fFeatures;
+   std::optional<IntFlag_t> fMaxOpenedFiles;
+   std::optional<IntFlag_t> fVerbosity;
+   std::optional<IntFlag_t> fCompressionSettings;
+
+   int fOutputArgIdx;
+   int fFirstInputIdx;
+};
+
+enum class EFlagResult { kIgnored, kParsed, kError };
```

Comment:
Careful about the redeclaration of `kError` (even though it's inside an enum class so it's unclear to me why the compiler cannot distinguish). Maybe an extra inline namespace would help?